notation.go

```diff
@@ -63,31 +45,207 @@ type SignOptions struct {
+// performed.
+type VerificationOutcome struct {
+	// RawSignature is the signature envelope blob
+	RawSignature []byte
+
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*ValidationResult
+
+	// Error that caused the verification to fail (if it fails)
+	Error error
 }
 
 // Verifier is a generic interface for verifying an artifact.
 type Verifier interface {
-	// Verify verifies the signature and returns the verified descriptor and
-	// metadata of the signed artifact.
-	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (Descriptor, error)
+	// Verify verifies the signature blob and returns the artifact
+	// descriptor upon successful verification.
+	// If nil signature is present and the verification level is not 'skip',
+	// an error will be returned.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if reflect.DeepEqual(outcome.VerificationLevel, trustpolicy.LevelSkip) {
+		return ocispec.Descriptor{}, []*VerificationOutcome{outcome}, nil
+	}
+	// get signature manifests
+	artifactRef := opts.ArtifactReference
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+
+	var verificationOutcomes []*VerificationOutcome
+	var targetArtifactDesc ocispec.Descriptor
+	if opts.MaxSignatureAttempts <= 0 {
+		// Set MaxVerificationLimit to 50 as default
+		opts.MaxSignatureAttempts = maxVerificationLimitDefault
+	}
+	errExceededMaxVerificationLimit := ErrorVerificationFailed{Msg: fmt.Sprintf("number of signatures associated with an artifact should be less than: %d", opts.MaxSignatureAttempts)}
+	count := 0
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		// process signatures
+		for _, sigManifestDesc := range signatureManifests {
```

Comment:
Callout: we haven't implemented signature filtering https://github.com/notaryproject/notaryproject/blob/main/specs/signature-specification.md#signature-filtering

Reply:
Opened issue: https://github.com/notaryproject/notation-go/issues/197